index.go

```diff
@@ -0,0 +1,171 @@
+type IndexOperation struct {
+	Operation IndexOp
+	Indices   []string
+	Config    *indexRequestConfig
+	Doc       RoutableDoc
+}
+
+func NewIndexCreateOperation(index string, doc RoutableDoc, ops ...IndexReqOptionFunc) *IndexOperation {
+	config := &indexRequestConfig{}
+	for _, o := range ops {
+		o(config)
+	}
+	return &IndexOperation{Indices: []string{index}, Doc: doc, Config: config, Operation: IndexCreate}
+}
+
+func NewIndexDeleteOperation(index []string, ops ...IndexReqOptionFunc) *IndexOperation {
+	config := &indexRequestConfig{}
+	for _, o := range ops {
+		o(config)
+	}
+	return &IndexOperation{Indices: index, Config: config, Operation: IndexCreate}
+}
+
+func NewIndexGetOperation(index []string, ops ...IndexReqOptionFunc) *IndexOperation {
+	config := &indexRequestConfig{}
+	for _, o := range ops {
+		o(config)
+	}
+	return &IndexOperation{Indices: index, Config: config, Operation: IndexGet}
+}
+
+func NewIndexExistOperation(index []string, ops ...IndexReqOptionFunc) *IndexOperation {
+	config := &indexRequestConfig{}
+	for _, o := range ops {
+		o(config)
+	}
+	return &IndexOperation{Indices: index, Operation: IndexExists}
+}
+func (i *IndexOperation) OperationType() IndexOp {
+	return i.Operation
+}
+
+func (i *IndexOperation) GetIndices() []string {
+	return i.Indices
+}
+
+func (i *IndexOperation) GetDoc() RoutableDoc {
+	return i.Doc
+}
+
+type indexRequestConfig struct {
+	masterTimeout         time.Duration
+	clusterManagerTimeout time.Duration
+	expandWildcards       string // make this enum later
+	waitForActiveShards   string
+	allowNoIndices        bool
+}
+
+// IndexReqOptionFunc for functional options
+type IndexReqOptionFunc func(config *indexRequestConfig)
+
+// WithMasterTimeOut to add the master timeout
+func WithMasterTimeOut(d time.Duration) IndexReqOptionFunc {
+	return func(config *indexRequestConfig) {
+		config.masterTimeout = d
+	}
+}
+
+// WithClusterManagerTimeout to add the cluster master timeout
+func WithClusterManagerTimeout(d time.Duration) IndexReqOptionFunc {
+	return func(config *indexRequestConfig) {
+		config.clusterManagerTimeout = d
+	}
+}
+
+// WithExpandWildcards to add the wildcard option
+func WithExpandWildcards(w string) IndexReqOptionFunc {
+	return func(config *indexRequestConfig) {
+		config.expandWildcards = w
+	}
+}
+
+// WithWaitForActiveShards to add the active shard option
+func WithWaitForActiveShards(s string) IndexReqOptionFunc {
+	return func(config *indexRequestConfig) {
+		config.waitForActiveShards = s
+	}
+}
+
+// WithAllowNoIndices to set allow no index option
+func WithAllowNoIndices(a bool) IndexReqOptionFunc {
+	return func(config *indexRequestConfig) {
+		config.allowNoIndices = a
+	}
+}
+
+// Validate validating things
+func (c *indexRequestConfig) Validate() error {
+	// implement some validation e.g. for active shard a positive value etc.
+	return nil
+}
+
+// IndexResponse - the combined response, may be create an interface and has each response its own sturct
```

Comment:
I think this indicates that these should be separate requests. Rather than one `IndexRequest` we should have `CreateIndexRequest`, `GetIndexRequest`, etc. I'm on the fence, but the body's and responses feel varied enough that I lean towards it.